### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/.DS_Store
Comment (PerryBaran) [inline src/.DS_Store:1]: this file shouldn't be here. From a bit of googling it looks like it is automatically created by Max OS X Finder in browsed directories. If this file contains something you require to make it work, then it should be added to the .gitignore, otherwise it should be deleted

File: src/port.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        this.ships = []
+    }
+    addShip(ship){
+        this.ships.push(ship)
+
+    }
+    removeShip(ship){
+        const index = this.ships.indexOf(ship)
+        this.ships.splice(index, 1)
+    }
```
Comment (PerryBaran) [inline src/port.js:13]: add an empty line between methods to improve readability

File: src/port.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class Port{
+    constructor (name) {
+        this.name = name
+        this.ships = []
+    }
+    addShip(ship){
+        this.ships.push(ship)
+
+    }
```
Comment (PerryBaran) [inline src/port.js:9]: remove empty line on line 8

File: src/ship.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+class Ship{
+    constructor(itinerary){
+    this.itinerary = itinerary
+    this.currentPort = itinerary.ports[0]
+    this.previousPort = null
+    this.currentPort.addShip(this)
```
Comment (PerryBaran) [inline src/ship.js:6]: these properties should be tabbed in by one

File: src/ship.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        this.currentPort = itinerary.ports[previousPortIndex + 1]
+    };
+    
+    setSail() {
+        const itinerary = this.itinerary;
+        const currentPortIndex = itinerary.ports.indexOf(this.currentPort)
+          
+        if (currentPortIndex === (itinerary.ports.length - 1)) {
+              throw new Error('Reached the end of the Itinerary!')
+            }
```
Comment (PerryBaran) [inline src/ship.js:22]: closing bracket on line 22 should be on the same level of indentation as if on line 20. A general rule is the closing bracket should be on the same indentation as the line that has the opening bracket

File: src/ship.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+          
+        this.currentPort.removeShip(this)
+        this.previousPort = this.currentPort
+        this.currentPort = null
+        }
```
Comment (PerryBaran) [inline src/ship.js:27]: closing bracket on line 27 should be on the same level of indentation as setSail

File: __test__/itinerary.test.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const Itinerary = require('../src/itinerary')
+const Port = require('../src/port.js')
+
+describe('itinerary constructor', () => {
+    it('it returns an object', () => {
+        
+        expect(new Itinerary('Itinerary')).toBeInstanceOf(Object);
```
Comment (PerryBaran) [inline __test__/itinerary.test.js:7]: remove empty line on line 6. Only have an empty line before the expect statement if you are doing some setup beforehand.

File: __test__/itinerary.test.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    });
+    it('it has the ports property', () => {
+        const amsterdam = new Port('Amsterdam');
+        const dover = new Port ('Dover')
+
+        const itinerary = new Itinerary([amsterdam, dover])
+
+        expect(itinerary.ports).toEqual([amsterdam, dover])
+    });
```
Comment (PerryBaran) [inline __test__/itinerary.test.js:16]: add an empty space between tests

File: __test__/itinerary.test.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        
+        expect(new Itinerary('Itinerary')).toBeInstanceOf(Object);
+    });
+    it('it has the ports property', () => {
+        const amsterdam = new Port('Amsterdam');
+        const dover = new Port ('Dover')
+
+        const itinerary = new Itinerary([amsterdam, dover])
+
+        expect(itinerary.ports).toEqual([amsterdam, dover])
```
Comment (PerryBaran) [inline __test__/itinerary.test.js:15]: You could use mock data to replace the ports. In this test, you are just checking that by passing an array to the itinerary it sets it to the ports, so you could refactor the test to look like
```
const ports = ['port1', 'port2']
const itinerary = new Itnerary(ports)
expect(itinerary.ports).toEqual(ports)
```

File: __test__/port.test.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+            const port = new Port('Amsterdam')
+            const theBlackPearl = {}
+            const theFlyingDutchman = {}
+
+            port.addShip(theBlackPearl)
+            port.addShip(theFlyingDutchman)
+            port.removeShip(theBlackPearl)
+    
+            expect(port.ships).toEqual([theBlackPearl])
```
Comment (PerryBaran) [inline __test__/port.test.js:39]: add empty lines between test cases, and an empty line after the beforeEach to improve readability

Comment (PerryBaran) [inline __test__/port.test.js:39]: remove empty lines at the top of the test cases. There shouldn't ever be an empty line immediately below the `it('...', () => {...` line

Comment (PerryBaran) [inline __test__/port.test.js:39]: You create `theBlackPeral` and `theFlyingDutchman` variables in the `beforeEach` but you only make use of them in the last test, where you create them again anyway. if you only need a variable for one test only create it for that test, i.e. remove them from the `beforeEach` and remove `let theBlackPearal` and `let theFlyingDutchman`

Comment (PerryBaran) [reply __test__/port.test.js:39]: Same as `ship`, as it is only used in the `'can add a ship'` test. Also, I'm surprised the test works, as you didn't initialise a variable for `ship` beforehand (i.e. `let ship`), but after some testing that doesn't seem to matter in this project and I don't know why. I tested this in some projects I made and they fail if I don't so I will have to look into that some more to see if I can figure out why, but I would suggest always initialises your variable beforehand

Comment (PerryBaran) [inline __test__/port.test.js:39]: You've already created a port in your `beforeEach`, so you could just use that port in the `'returns an object'` and '`can remove a ship'` tests, instead of calling `new Port` seperately in those tests

File: __test__/port.test.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+const Port = require('../src/port.js')
+
+describe('port constructor', () => {
+    describe('has ports and ships', () => {
```
Comment (PerryBaran) [inline __test__/port.test.js:4]: Don't need to put these tests inside 2 describe blocks. Describe blocks are used to separate out different test cases, but as all your tests are inside of both of these you only need 1 describe block

File: __test__/ship.test.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        const dover = new Port ('Dover')
+        const itinerary = new Itinerary ([amsterdam, dover])
+        const ship = new Ship (itinerary)
+
+        ship.setSail()
+        ship.dock()
+
+        expect(() => ship.setSail()).toThrowError('Reached the end of the Itinerary!')
+    })
+})
```
Comment (PerryBaran) [inline __test__/ship.test.js:83]: more indentation and formatting issues here

File: __test__/ship.test.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        it('it should have a starting port', () => {
+
+             expect(ship.currentPort).toBe(amsterdam)
+        })
+
+        it('can set sail', () => {
+    
+            ship.setSail()
+          
+            expect(ship.currentPort).toBeFalsy()
```
Comment (PerryBaran) [inline __test__/ship.test.js:49]: I would have written this test to be `expect(ship.currentPort).toEqual(null)` as it's more explicit with the value currentPort is being set to

File: __test__/ship.test.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
+const Ship = require('../src/ship')
+const Port = require('../src/port')
+const Itinerary = require('../src/itinerary')
```
Comment (PerryBaran) [inline __test__/ship.test.js:3]: These tests are for testing the `Ship` constructor. I would recommend trying to use mock data to replace `Port` and `Itinerary`. You've already done this for most tests in this file, you only need to change it for the last test

File: __test__/ship.test.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+          })
+
+          it('can dock at a different port', () => {
+
+              ship.setSail()
+              ship.dock()
+              
+              expect(ship.currentPort).toBe(dover)
+              expect(amsterdam.removeShip).toHaveBeenCalledWith(ship)
+        })
```
Comment (PerryBaran) [inline __test__/ship.test.js:65]: Good use of test doubles

File: __test__/ship.test.js (hunk 11 above)
Comment (PerryBaran) [inline __test__/ship.test.js:83]: Why is this test on it's own? it doesn't need it's own describe block and the test should be moved to the other describe block. Above, you already have 2 nested describe blocks. If you move this test inside of the 1st describe block but outside of the 2nd e.g.
```
describe('ship', () => {
  describe('has ports and itinerarys', () => {
    ...
  }

  it('can\'t sail further than the itinerary', () => {
    ...
  }
})
```
you could give this test separate setup to the rest of the tests if that is what you'd want

File: package.json [hunk 15]
```diff
@@ -4,11 +4,15 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "test": "jest"
+    "test": "jest",
+    "format": "npx prettier --write ."
```
Comment (PerryBaran) [inline package.json:8]: I like that you've added .prettierrc rules, and added this command to your package.json to format your code. If you ran this command, `npm run format`, before submitting this code review it would have fixed most of the problems you have with formatting

**Review by PerryBaran — COMMENTED** [review]
Lots of good stuff, code works as expected. However, there is lots of problems with formatting, which would be easily fixed if you ran the command that you have added to your package.json to format your code with prettier i.e. run in your terminal `npm run format`. Also, I would add a README that includes some details about what this project is, why you made this project, how this project works and gives attribution.
